### PR TITLE
Use #[expect(…)] instead of #[allow(…)], remove unnecessary allows

### PR DESCRIPTION
**Pull request by riesentoaster**
[`#[expect(…)]`](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-expect-attribute) returns a lint warning (configured to be an error) if the lint warning is not present without the attribute.

This also means I was able to remove *a lot* of `allow` statements.

I suggest in the future, `expect` is used wherever possible. The only exceptions I've stumbled across are when they are used in a macro and only appear for some macro calls (think `trivial_numeric_casts` of a `u64` to a type passed to the struct, and it is called for `u64` among others) and if the warning only appears in certain cfgs.

### CODE REVIEW

**Comment by domenukk** [comment]
Feel free to fix this for all build targets - but it might be a world of pain :)

**Comment by riesentoaster** [comment]
Also: I cannot guarantee that I haven't broken anything. I've manually checked as much as possible and the CI found stuff as well, but I may have still missed some cfg-dependent stuff (there's a lot in there lol). On the upside, I've only touched linting-related things. So if anyone's computer starts complaining, it's easy to just re-add the warning or change it back from an `expect` to an `allow`.

File: fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml [hunk 1]
```diff
@@ -111,7 +111,10 @@ windows_alias = "unsupported"
 script_runner = "@shell"
 script = '''
 rm -rf libafl_unix_shmem_server || true
-timeout 31s ./${FUZZER_NAME} --cores 0 --input ./corpus | tee fuzz_stdout.log 2>/dev/null || true
+
+# start manager in background, then start one client
+timeout 32s ./${FUZZER_NAME} --cores 0 --input ./corpus 2>/dev/null &
+timeout 31s ./${FUZZER_NAME} --cores 1 --input ./corpus | tee fuzz_stdout.log 2>/dev/null || true
```
Comment (riesentoaster) [inline fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml:117]: I added this, otherwise the test doesn't pass. Doesn't either on main btw.

**Comment by riesentoaster** [comment]
Not quite sure why I did this. But it's done.

File: fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml (hunk 1 above)
Comment (domenukk) [reply fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml:117]: Weird, how was this not caught before?

File: fuzzers/inprocess/libfuzzer_libpng_accounting/src/lib.rs [hunk 2]
```diff
@@ -21,246 +18,197 @@ use libafl::{
...
-}
-
-/// The main fn, `no_mangle` as it is a C symbol
+/// The main fn, `no_mangle` as it is a C main
 #[no_mangle]
 pub extern "C" fn libafl_main() {
     // Registry the metadata types used in this fuzzer
     // Needed only on no_std
     // unsafe { RegistryBuilder::register::<Tokens>(); }
-    let opt = Opt::parse();
```
Comment (domenukk) [inline fuzzers/inprocess/libfuzzer_libpng_accounting/src/lib.rs:126]: Why did you remove opt parsing?

File: fuzzers/inprocess/libfuzzer_libpng_accounting/src/lib.rs [hunk 3]
```diff
@@ -21,246 +18,197 @@ use libafl::{
...
-}
-
-/// The commandline args this fuzzer accepts
-#[derive(Debug, Parser)]
-#[command(
-    name = "libfuzzer_libpng_launcher",
-    about = "A libfuzzer-like fuzzer for libpng with llmp-multithreading support and a launcher",
-    author = "Andrea Fioraldi <andreafioraldi@gmail.com>, Dominik Maier <domenukk@gmail.com>"
-)]
-struct Opt {
```
Comment (domenukk) [inline fuzzers/inprocess/libfuzzer_libpng_accounting/src/lib.rs:57]: This shouldn't be removed

File: fuzzers/structure_aware/baby_fuzzer_gramatron/src/main.rs [hunk 4]
```diff
@@ -26,7 +26,7 @@ use libafl::{
 };
 use libafl_bolts::{rands::StdRand, tuples::tuple_list};
 
-/// Coverage map with explicit assignments due to the lack of instrumentation
+/// Coverage map with explicit assignments due to the lack of instrumentationnly a problem in nightly
```
Comment (domenukk) [inline fuzzers/structure_aware/baby_fuzzer_gramatron/src/main.rs:29]: Comment is .. weird.

File: fuzzers/structure_aware/baby_fuzzer_grimoire/src/main.rs [hunk 5]
```diff
@@ -83,6 +83,7 @@ pub fn main() {
     };
 
     // Create an observation channel using the signals map
+    #[allow(static_mut_refs)] // only a problem in nightly
```
Comment (domenukk) [inline fuzzers/structure_aware/baby_fuzzer_grimoire/src/main.rs:86]: We should not use static mut refs here. 
See this discussion as reference
https://github.com/AFLplusplus/LibAFL/pull/2770#discussion_r1885707774

File: fuzzers/structure_aware/baby_fuzzer_tokens/src/main.rs [hunk 6]
```diff
@@ -20,6 +20,7 @@ use libafl_bolts::{rands::StdRand, tuples::tuple_list};
 
 /// Coverage map with explicit assignments due to the lack of instrumentation
 static mut SIGNALS: [u8; 16] = [0; 16];
+#[allow(static_mut_refs)] // only a problem in nightly
```
Comment (domenukk) [inline fuzzers/structure_aware/baby_fuzzer_tokens/src/main.rs:23]: Don't allow static mut refs

File: fuzzers/structure_aware/baby_fuzzer_tokens/src/main.rs [hunk 7]
```diff
@@ -67,6 +67,7 @@ pub fn main() {
     };
 
     // Create an observation channel using the signals map
+    #[allow(static_mut_refs)] // only a problem in nightly
```
Comment (domenukk) [inline fuzzers/structure_aware/baby_fuzzer_tokens/src/main.rs:70]: Nope

File: fuzzers/structure_aware/nautilus_sync/src/bin/libafl_cc.rs [hunk 8]
```diff
@@ -2,7 +2,6 @@ use std::env;
 
 use libafl_cc::{ClangWrapper, CompilerWrapper, ToolWrapper};
 
-#[allow(clippy::missing_panics_doc)]
```
Comment (domenukk) [inline fuzzers/structure_aware/nautilus_sync/src/bin/libafl_cc.rs:5]: pretty sure this is needed you build this with `clippy -- -D clippy::pedantic` (which I like to do)
Anyway it doesn't hurt right?

File: libafl/src/common/nautilus/regex_mutator/mod.rs [hunk 9]
```diff
@@ -53,7 +53,6 @@ fn append_unicode_range<R: Rand>(
     cls: ClassUnicodeRange,
 ) {
     let mut chr_a_buf = [0; 4];
-    #[allow(clippy::similar_names)]
```
Comment (domenukk) [inline libafl/src/common/nautilus/regex_mutator/mod.rs:56]: Pretty sure this one is a thing with clippy::pedantic

File: libafl/src/executors/hooks/inprocess.rs [hunk 10]
```diff
@@ -197,8 +197,7 @@ where
 {
     fn init<E: HasObservers>(&mut self, _state: &mut S) {}
     /// Call before running a target.
-    #[allow(clippy::unused_self)]
```
Comment (domenukk) [inline libafl/src/executors/hooks/inprocess.rs:200]: `unused_self` will happen in apple + no_std

File: libafl/src/executors/hooks/timer.rs [hunk 11]
```diff
@@ -201,8 +201,6 @@ impl TimerStruct {
 
     #[cfg(target_os = "linux")]
     #[must_use]
-    #[allow(unused_unsafe)]
```
Comment (domenukk) [inline libafl/src/executors/hooks/timer.rs:204]: `#[cfg(not(miri))]` -> this `unused_unsafe` for miri

File: libafl/src/executors/hooks/timer.rs [hunk 12]
```diff
@@ -324,7 +322,6 @@ impl TimerStruct {
 
     /// Disable the timer
     #[cfg(target_os = "linux")]
-    #[allow(unused_variables)]
```
Comment (domenukk) [inline libafl/src/executors/hooks/timer.rs:327]: see `#[cfg(not(miri))]`

**Comment by domenukk** [comment]
This is cool! I'm a bit afraid we'll run into a bunch of new errors on odd platforms though

File: fuzzers/inprocess/libfuzzer_libpng_accounting/src/lib.rs (hunk 2 above)
Comment (riesentoaster) [reply fuzzers/inprocess/libfuzzer_libpng_accounting/src/lib.rs:126]: Oh, …, ehm, …, that shouldn't have been removed.

Not sure what happened there. I guess you shouldn't program late at night.

**Comment by riesentoaster** [comment]
1. Dead code/unused functions: I have decided to not remove anything like that for now, this PR is large enough by itself. Separate PR?
2. `static_mut_refs`: Same thing. Should be fairly easy to do, since you can basically grep for it. It may be worth unifying the pseudo-coverage across example fuzzers anyways. There is a variance in naming conventions, how the observers are built, etc. And that seems odd to me.
3. Pedantic: I'll fix what breaks. If you want it in the repo: Should we not add this to the CI as well?

**Comment by domenukk** [comment]
- static_mut_refs: I don't think we should allow them, then we'll never fix them (and IIRC they will break in rust 2024)
- Dead code/unused functions: same thing, we shound't allow those?

**Comment by domenukk** [comment]
The pendantic comment you can ignore

**Comment by riesentoaster** [comment]
Both are already in the repo. This makes them just a bit more visible.

I need to do something else at some point. Separate PR please?

**Comment by domenukk** [comment]
Ah you already didn't :D 
It's fine if we have a good amount of clippy lints

**Comment by domenukk** [comment]
As in: the maximum amount of enabled lint possible at all times; without being ridicoulous

File: libafl/src/executors/hooks/inprocess.rs (hunk 10 above)
Comment (riesentoaster) [reply libafl/src/executors/hooks/inprocess.rs:200]: It does not :)

(At least not on my computer, nor in the CI.)

Comment (domenukk) [reply libafl/src/executors/hooks/inprocess.rs:200]: Strange, then the lint changed

File: fuzzers/baby/baby_fuzzer/src/main.rs [hunk 13]
```diff
@@ -24,14 +24,15 @@ use libafl_bolts::{current_nanos, nonzero, rands::StdRand, tuples::tuple_list, A
 
 /// Coverage map with explicit assignments due to the lack of instrumentation
 static mut SIGNALS: [u8; 16] = [0; 16];
+#[allow(static_mut_refs)] // only a problem in nightly
```
Comment (domenukk) [inline fuzzers/baby/baby_fuzzer/src/main.rs:27]: In any case these allows have to go

File: fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml (hunk 1 above)
Comment (domenukk) [reply fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml:117]: Does it still not work now that you fixed the Opt parser btw?

File: fuzzers/baby/baby_fuzzer/src/main.rs (hunk 13 above)
Comment (domenukk) [reply fuzzers/baby/baby_fuzzer/src/main.rs:27]: Or otherwise leave a biiiiiig TODO comment

Comment (riesentoaster) [reply fuzzers/baby/baby_fuzzer/src/main.rs:28]: I've created an issue (#2786) and added a comment to the code with a link to it.

File: fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml (hunk 1 above)
Comment (riesentoaster) [reply fuzzers/inprocess/libfuzzer_libpng_accounting/Makefile.toml]: I've reverted the change. Most likely something to do with the unwanted code changes. No idea what happened there lol.

**Comment by riesentoaster** [comment]
> Ah you already didn't :D

Who needs to graduate anyways…

**Comment by domenukk** [comment]
@devnexen if you ever find the time, it'd be interesting if this breaks something on obscure *Nixes

**Comment by devnexen** [comment]
Ok, will do !

**Comment by devnexen** [comment]
starting with solaris/illumos, nothing worrying but I get a bunch of (jusitifed) warnings about the new `trivial_numeric_cast` lint when calling ioctl because the op argument is an int on these platforms.

**Comment by domenukk** [comment]
Thank you @devnexen ! :)
